qctrlopencontrols/pyquil/program.py

```diff
@@ -0,0 +1,209 @@
+        raise ArgumentsValueError('Dynamical decoupling sequence is not recognized.'
+                                  'Expected DynamicDecouplingSequence instance',
+                                  {'type(dynamic_decoupling_sequence)':
+                                       type(dynamic_decoupling_sequence)})
+
+    if target_qubits is None:
+        target_qubits = [0]
+
+    if gate_time <= 0:
+        raise ArgumentsValueError(
+            'Time delay of identity gate must be greater than zero.',
+            {'gate_time': gate_time})
+
+    if np.any(target_qubits) < 0:
+        raise ArgumentsValueError(
+            'Every target qubits index must be positive.',
+            {'target_qubits': target_qubits})
+
+    if algorithm not in [FIX_DURATION_UNITARY, INSTANT_UNITARY]:
+        raise ArgumentsValueError('Algorithm must be one of {} or {}'.format(
+            INSTANT_UNITARY, FIX_DURATION_UNITARY), {'algorithm': algorithm})
+
+    unitary_time = 0.
+    if algorithm == FIX_DURATION_UNITARY:
+        unitary_time = gate_time
+
+    rabi_rotations = dynamic_decoupling_sequence.rabi_rotations
+    azimuthal_angles = dynamic_decoupling_sequence.azimuthal_angles
+    detuning_rotations = dynamic_decoupling_sequence.detuning_rotations
+
+    if len(rabi_rotations.shape) == 1:
+        rabi_rotations = rabi_rotations[np.newaxis, :]
+    if len(azimuthal_angles.shape) == 1:
+        azimuthal_angles = azimuthal_angles[np.newaxis, :]
+    if len(detuning_rotations.shape) == 1:
+        detuning_rotations = detuning_rotations[np.newaxis, :]
+
+    operations = np.vstack((rabi_rotations, azimuthal_angles, detuning_rotations))
+    offsets = dynamic_decoupling_sequence.offsets
+
+    time_covered = 0
+    program = Program()
+    program += Pragma('PRESERVE_BLOCK')
+
+    for operation_idx in range(operations.shape[1]):
+
+        offset_distance = offsets[operation_idx] - time_covered
+
+        if np.isclose(offset_distance, 0.0):
+            offset_distance = 0.0
+
+        if offset_distance < 0:
+            raise ArgumentsValueError("Offsets cannot be placed properly",
+                                      {'sequence_operations': operations})
+
+        if offset_distance > 0:
+            while (time_covered+gate_time) <= offsets[operation_idx]:
+                for qubit in target_qubits:
+                    program += I(qubit)
+                time_covered += gate_time
+
+        rabi_rotation = operations[0, operation_idx]
+        azimuthal_angle = operations[1, operation_idx]
+        x_rotation = rabi_rotation * np.cos(azimuthal_angle)
+        y_rotation = rabi_rotation * np.sin(azimuthal_angle)
+        z_rotation = operations[2, operation_idx]
+
+        rotations = np.array([x_rotation, y_rotation, z_rotation])
+        zero_pulses = np.isclose(rotations, 0.0).astype(np.int)
+        nonzero_pulse_counts = 3 - np.sum(zero_pulses)
+        if nonzero_pulse_counts > 1:
+            raise ArgumentsValueError(
```

Comment:
Again, this isn't really an ArgumentsValueError is it? In this case I think it means *we* messed something up, because we produced an invalid DDS?

Reply:
Again, Open-Controls only has `ArgumentsValueError`

Reply:
See before.

Specifically here, the offset, rabi_rotation, azimuthal_angle and detuning rotation should all be in the extras dictionary.